### DISCUSSION TITLE
How to 'conditionally' support a capability

### DISCUSSION THREAD

**Discussion by andyw8** [microsoft/vscode-discussions#561] [Extension Development QnA] [answered]
:wave: I’m looking for some general guidance on how to ‘conditionally’ support a capability.
I have an extension which implements formatting. It supports two different formatting libraries. If the codebase has a dependency on one of these, formatting should be available.
On the server side, I could use design it such that `textDocument/formatting` is conditional on the availability of a formatter. That means if someone tries to use the Format feature when none is available, they would see an alert like the one below. But would be highly confusing to the user, because it suggests they need to install a different extension.

Instead, my intention is have a StatusItem that indicates that formatting is not available, which a clickable link that gives further explanation.

The part I’m unclear on is how best to communicate that from the server to the client. I’m thinking it would either be part of the initialization process, or some kind of custom notification.

Is there guidance or recommendations on an approach for this?

<img width="278" alt="Screenshot 2023-04-10 at 2 06 41 PM" src="https://user-images.githubusercontent.com/13400/231216023-1d580e51-1cdf-4ae7-b0f8-f87a053a6115.png">

**Comment by r3m0t** [comment, marked as answer]
You mention client and server so I guess you're using LSP?

I would  register your ability to format unconditionally and `throw new Error("message")` if the dependency is missing. If the error shown is not good then subclass `LanguageClient` and override `handleFailedRequest` to detect the specific error and show a better notification.
